### ISSUE TITLE
Exception 400, message='unexpected content-length header' Error

### ISSUE THREAD

**Issue by garyak** [PeteRager/lennoxs30#107]
Seeing these repeating error messages (occasionally):

`2021-11-23 23:25:28 ERROR (MainThread) [custom_components.lennoxs30] messagePump_task - host [192.168.3.94] S30Exception (S30Exception(...), "messagePump Failed - Exception 400, message='unexpected content-length header', url=URL('https://192.168.3.94/Messages/ha_prod/Retrieve?Direction=Oldest-to-Newest&MessageCount=10&StartTime=1&LongPollingTimeout=15')")`

`2021-11-24 03:59:59 ERROR (MainThread) [custom_components.lennoxs30] messagePump_task - host [192.168.3.76] S30Exception (S30Exception(...), 'messagePump Failed - Exception Server disconnected')`

sysUpTime: 3709140
softwareVersion: 3.81.207

s30's remain connected and operational during these events. Probably see a series of these once or twice a day. I'm not running the cloud version concurrently.

**Comment by PeteRager** [comment]
Ok.  Good information.  We are both running the same firmware version.   Your system has been up for 42 days.   We did have a report of the S30 rebooting on local connections with the first alpha, but it hasn’t reoccurred since.  You are not having that problem.  I had some issue when running both the cloud and local connections - which also are no longer reproducing.   You are not doing that.

It may be these two errors are just going to happen - just like in the cloud when we periodically were getting 502 errors - and the remedy will be not to log them as errors.

For now, let’s keep an eye on it.  I will open an issue, so that if other users are seeing it also they can provide information.  

Pete

**Comment by PeteRager** [comment]
I have been researching this.  There are some known issues with the aiohttp library the integration is using which causes it to reject http messages that do not comply to the “standard”.  Since not every web server is perfect, this strict interpretation of the standard causes issues. 

https://github.com/aio-libs/aiohttp/issues/2062

So I believe that is what is happening here.   The first step, is To some additional logging, to see what is in the “non standards compliant” message, so we can determine the action to take.  If it’s an empty message, the we can safely ignore it, if it’s a message with data - then we’d need a different strategy.

Right now I am trying to reproduce.  I had seen this error a month ago, but have not seen it since.  I have updated my HA system so it is possible the error is specific to a version of the aiohttp library. 

@garyak - what Home Assistant Release are you running? Approximately how often does the error show up?

**Comment by PeteRager** [comment]
On my development system I have aiohttp=3.8.0
On my prod system I have aiohttp=3.7.4.post0
On my test system I have aiohttp=3.7.4.post0
3.7.4.post0 was adopted by HA 10 months ago.

**Comment by garyak** [comment]
core-2021.11.5
Home Assistant OS 6.6
supervisor-2021.10.8

Recently I've seen these appear once or twice daily. However, nothing logged yet for today. Latest:

```
2021-11-25 03:05:33 ERROR (MainThread) [lennoxs30api.s30api_async] messagePump Failed - Exception Server disconnected
2021-11-25 03:05:33 ERROR (MainThread) [custom_components.lennoxs30] messagePump_task - host [192.168.3.76] S30Exception (S30Exception(...), 'messagePump Failed - Exception Server disconnected')
```

**Comment by PeteRager** [comment]
Thanks - core-2021.11.5 has aiohttp==3.7.4.post0

So I am running a different version in Dev; which is also where I'm testing; which may be masking the issue.

Yes, the ServerDisconnect error would happen if there is a network disruption.  That one I will update to be an Info message and track it in the metrics.

**Comment by PeteRager** [comment]
In aiohttp 3.8.0 which will be in the next HA release, the http parsing code has been replaced with a new library.  The new library appears to be less strict in processing headers.  Hence, it's possible this error will not longer occur once HA move to the new version.  This also may explain why I saw this error early on and once I rebased by Dev system to the current HA code-base I no longer saw the error.

So I think the approach will be this:

1. Add additional log information when the "unexpected content-length" error is encountered.
2. Handle the "Service Disconnected" error more gracefully (info level, not error)
3. Put this into the next release LennoxS30 release - early / mid december

Certainly if the error becomes more frequent and/or starts impacting operation of the Thermostat - we can revisit this plan - and release items 1+2 sooner to diagnose / remedy.

Thoughts?


Thoughts?

**Comment by garyak** [comment]
The errors don't appear to effect normal operation. I haven't encountered a UI failure. Wrapping this all up in a mid-December release is soon enough. After 2021.12.0 releases, I'll keep an eye on the log and let you know if the errors have resolved.

**Comment by PeteRager** [comment]
I have completed the changes for handling this error better and providing more diagnostic information so we can see exactly what is going on.  

I have also started working on moving the configuration into the UI, so we’ll have the ability to manage the configuration through the integration panel and have stuff organized by devices.

Anyways, I could push the error handling improvements as a release this week.  Or wait until I have the config changes done also - likely towards end of December.

Thoughts?   Errors still occurring regularly?

**Comment by garyak** [comment]
I'm still OK waiting. Errors have become less frequent.

**Comment by PeteRager** [comment]
Was automatically closed as changes are made and merged into the upcoming release available by end of year.

**Comment by PeteRager** [comment]
I just saw this error occur on my system last night - running the 0.1.3 release.  It seemed to be tied into an S30 reboot.

2021-12-24 22:06:56 ERROR (MainThread) [lennoxs30api.s30api_async] messagePump Failed - Exception 400, message="Content-Length can't be present with Transfer-Encoding", url=URL('https://192.168.1.175/Messages/ha_prod/Retrieve?Direction=Oldest-to-Newest&MessageCount=10&StartTime=1&LongPollingTimeout=15')
2021-12-24 22:06:56 ERROR (MainThread) [custom_components.lennoxs30] messagePump_task - host [192.168.1.175] S30Exception (S30Exception(...), 'messagePump Failed - Exception 400, message="Content-Length can\'t be present with Transfer-Encoding", url=URL(\'https://192.168.1.175/Messages/ha_prod/Retrieve?Direction=Oldest-to-Newest&MessageCount=10&StartTime=1&LongPollingTimeout=15\')')
2021-12-24 22:06:58 ERROR (MainThread) [lennoxs30api.s30api_async] messagePump Failed - Exception 400, message="Content-Length can't be present with Transfer-Encoding", url=URL('https://192.168.1.175/Messages/ha_prod/Retrieve?Direction=Oldest-to-Newest&MessageCount=10&StartTime=1&LongPollingTimeout=15')
2021-12-24 22:06:58 ERROR (MainThread) [custom_components.lennoxs30] messagePump_task - host [192.168.1.175] S30Exception (S30Exception(...), 'messagePump Failed - Exception 400, message="Content-Length can\'t be present with Transfer-Encoding", url=URL(\'https://192.168.1.175/Messages/ha_prod/Retrieve?Direction=Oldest-to-Newest&MessageCount=10&StartTime=1&LongPollingTimeout=15\')')
2021-12-24 22:07:01 WARNING (MainThread) [lennoxs30api.s30api_async] S30 has rebooted sysId [LCC] old uptime [3745440] new uptime [180]

**Comment by garyak** [comment]
I don't typically log warnings, so would miss the correlation to a thermostat reboot. I'll turn warnings back on and keep an eye on it.

**Comment by PeteRager** [comment]
This release has improved diagnostic logging for when that error occurs.  Also the Config is now in the integration UI.

https://github.com/PeteRager/lennoxs30/releases/tag/v0.2.0

**Comment by garyak** [comment]
I'll update to 0.2.0 and let you know how it goes.

**Comment by garyak** [comment]
I've updated to 0.2.1rc1. I'll keep an eye on it.

**Comment by garyak** [comment]
Caught this:

```
2022-01-26 17:35:59 ERROR (MainThread) [lennoxs30api.s30api_async] messagePump exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/lennoxs30api/s30api_async.py", line 555, in messagePump
    resp = await self.get(url, headers=headers, params=params)
  File "/usr/local/lib/python3.9/site-packages/lennoxs30api/s30api_async.py", line 353, in get
    resp = await self._session.get(
  File "/usr/local/lib/python3.9/site-packages/aiohttp/client.py", line 559, in _request
    await resp.start(conn)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/client_reqrep.py", line 913, in start
    self._continue = None
  File "/usr/local/lib/python3.9/site-packages/aiohttp/helpers.py", line 721, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
2022-01-26 17:35:59 ERROR (MainThread) [custom_components.lennoxs30] messagePump_task - general error host [192.168.3.94] Code [5] Reference [5] [messagePump exception]
2022-01-26 17:36:00 ERROR (MainThread) [lennoxs30api.s30api_async] MessagePump clientResponse please report this information as an Issue - Error while executing request: [Content-Length can't be present with Transfer-Encoding]: error=<class 'aiohttp.client_exceptions.ClientResponseError'>, resp.status=[400], resp.request_info=[RequestInfo(url=URL('https://192.168.3.94/Messages/ha_prod/Retrieve?Direction=Oldest-to-Newest&MessageCount=10&StartTime=1&LongPollingTimeout=15'), method='GET', headers=<CIMultiDictProxy('Host': '192.168.3.94', 'Accept': '*/*', 'Accept-Encoding': 'gzip, deflate', 'User-Agent': 'Python/3.9 aiohttp/3.8.1')>, real_url=URL('https://192.168.3.94/Messages/ha_prod/Retrieve?Direction=Oldest-to-Newest&MessageCount=10&StartTime=1&LongPollingTimeout=15'))], resp.headers=[None]
```
Error repeated for two seconds, then stopped. I use the iOS Companion App to interact with the s30s. Occasionally, I need to restart the App to regain control. Not sure if this is related or not. I'm on all the latest versions.

**Comment by PeteRager** [comment]
Thanks for the report.  I’ll review the errors.

Regarding iOS app, I periodically see it come up with a blank / disconnected screen and have to close it and reopen it.  Is that what you see?  This appears to be the same problem as the integration had when using cloud connection - being able to reliably get the config data.

**Comment by garyak** [comment]
The integration appears connected, but set point value remains red "pending" when changed. The s30 doesn't adjust to the new set point. If I restart just the App, set point adjustments are received by the s30.

**Comment by PeteRager** [comment]
Just to be clear, by app you mean the IOS app?

**Comment by garyak** [comment]
Yes.

**Comment by PeteRager** [comment]
I know their IOS app when on the LAN/WiFi goes direct to the thermostat.  When it’s not on the LAN it goes via the cloud connection.  The timeout error in the log you sent does indicate the thermostat was unreachable for the default timeout of 10 minutes.   I saw a similar error on mine a few weeks back.  Certainly this could occur if the WiFi was down, or the S30 reboots.  I suspect, it may also occur sometimes if the internet connection goes away

I plan on doing some controlled testing around this.

**Comment by garyak** [comment]
Loss of network connectivity is certainly a possibility. Breaking in a new Ubiquity network here. Four months and still wrestling with "issues".

**Comment by PeteRager** [comment]
Through a little bit of luck, I've been able to make this issue occur every 24 hours.  Turns out, this message means that the S30 has discarded the subscription and hence the components needs to re-initialize the connection.  So I have those changes in place, have cleaned up the error logs.

It started to repro when I put pfSense in and blocked all internet access for the S30.  Now every 24 hours at 6:40 AM I get this error.
